### PR TITLE
Natspec: Output "type" and "version" in documentation

### LINKED ISSUE #8922: Natspec version/kind
It is currently not easy to tell apart userdoc/devdoc and hard to be backward/forward compatible because there is no notion of versioning.

How about introducing two meta fields: `@version` and `@kind`. Since all tags start with non-@ there should be no clash.

Alternatively use the hashmark (#) as a special sign.

`@version` would be 1 for the current version of Natspec.
`@kind` would be `user` for "userdoc" and `developer" for "devdoc".

### PR DESCRIPTION
**Pull request by Marenz**
fixes #8922

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `b950435380e15b9767345d662b3b6464c813cc04`:
```
Coding style error:
 libsolidity/interface/Natspec.h:43: static const size_t NATSPEC_VERSION = 1;

```
Please check that your changes are working as intended.

File: Changelog.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ Language Features:
 
 Compiler Features:
  * Yul: Raise warning for switch statements that only have a default and no other cases.
+ * Natspec: Add fields "kind" and "version" in the JSON output.
```
Comment (leonardoalt) [inline Changelog.md]:
```suggestion
 * Natspec: Add fields "kind" and "version" to the JSON output.
```

File: libsolidity/interface/Natspec.cpp [hunk 2]
```diff
@@ -37,6 +37,9 @@ Json::Value Natspec::userDocumentation(ContractDefinition const& _contractDef)
 	Json::Value doc;
 	Json::Value methods(Json::objectValue);
 
+	doc["version"] = Json::Value(NATSPEC_VERSION);
+	doc["kind"]    = Json::Value("user");
```
Comment (leonardoalt) [inline libsolidity/interface/Natspec.cpp:41]: Is this our style? I would guess we don't use the extra spaces

File: libsolidity/interface/Natspec.h [hunk 3]
```diff
@@ -40,6 +40,8 @@ struct DocTag;
 class Natspec
 {
 public:
+	static size_t const NATSPEC_VERSION = 1;
```
Comment (leonardoalt) [inline libsolidity/interface/Natspec.h]:
```suggestion
	static size_t constexpr NATSPEC_VERSION = 1;
```

**Review by leonardoalt — COMMENTED** [review]
A few comments

File: libsolidity/interface/Natspec.cpp (hunk 2 above)
Comment (Marenz) [reply libsolidity/interface/Natspec.cpp:41]: Ah, no it isn't. Old habit. Will fix.

File: libsolidity/interface/Natspec.h [hunk 4]
```diff
@@ -40,6 +40,8 @@ struct DocTag;
 class Natspec
 {
 public:
+	static size_t constexpr NATSPEC_VERSION = 1;
```
Comment (chriseth) [inline libsolidity/interface/Natspec.h]: Don't we use this style only for preprocessor macros?

Comment (Marenz) [reply libsolidity/interface/Natspec.h]: Are you referring to the `constexpr` part or the `NATSPEC_VERSION` all-uppercase part? The former was from @leonardoalt suggestion, the later from our [style guide regarding constants](https://solidity.readthedocs.io/en/latest/style-guide.html#constants)

File: Changelog.md [hunk 5]
```diff
@@ -6,6 +6,7 @@ Language Features:
 Compiler Features:
  * Yul: Raise warning for switch statements that only have a default and no other cases.
  * Output compilation error codes.
+ * Natspec: Add fields "kind" and "version" to the JSON output.
```
Comment (hrkrshnn) [inline Changelog.md]:
```suggestion
 * NatSpec: Add fields "kind" and "version" to the JSON output.
```

For consistency with rest of changelog

File: libsolidity/interface/Natspec.cpp [hunk 6]
```diff
@@ -37,6 +37,9 @@ Json::Value Natspec::userDocumentation(ContractDefinition const& _contractDef)
 	Json::Value doc;
 	Json::Value methods(Json::objectValue);
 
+	doc["version"] = Json::Value(static_cast<Json::Value::UInt>(NATSPEC_VERSION));
```
Comment (hrkrshnn) [inline libsolidity/interface/Natspec.cpp]: If you make `NATSPEC_VERSION` as `unsigned int`, then we can avoid this static cast.

File: test/libsolidity/StandardCompiler.cpp [hunk 7]
```diff
@@ -354,9 +354,9 @@ BOOST_AUTO_TEST_CASE(basic_compilation)
 	BOOST_CHECK(contract["abi"].isArray());
 	BOOST_CHECK_EQUAL(util::jsonCompactPrint(contract["abi"]), "[]");
 	BOOST_CHECK(contract["devdoc"].isObject());
-	BOOST_CHECK_EQUAL(util::jsonCompactPrint(contract["devdoc"]), "{\"methods\":{}}");
+	BOOST_CHECK_EQUAL(util::jsonCompactPrint(contract["devdoc"]), "{\"kind\":\"dev\",\"methods\":{},\"version\":1}");
```
Comment (hrkrshnn) [inline test/libsolidity/StandardCompiler.cpp]:
```suggestion
	BOOST_CHECK_EQUAL(util::jsonCompactPrint(contract["devdoc"]), R"({"kind":"dev","methods":{},"version":1})");
```

File: test/libsolidity/StandardCompiler.cpp [hunk 8]
```diff
@@ -354,9 +354,9 @@ BOOST_AUTO_TEST_CASE(basic_compilation)
...
 	BOOST_CHECK(contract["userdoc"].isObject());
-	BOOST_CHECK_EQUAL(util::jsonCompactPrint(contract["userdoc"]), "{\"methods\":{}}");
+	BOOST_CHECK_EQUAL(util::jsonCompactPrint(contract["userdoc"]), "{\"kind\":\"user\",\"methods\":{},\"version\":1}");
```
Comment (hrkrshnn) [inline test/libsolidity/StandardCompiler.cpp]: Same here.

File: libsolidity/interface/Natspec.h (hunk 4 above)
Comment (hrkrshnn) [reply libsolidity/interface/Natspec.h]: https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md#4-variable-prefixes

I think it says prefix should be `s_`. Maybe camel case after that?

Comment (Marenz) [reply libsolidity/interface/Natspec.h]: It says `_s` for non-const, non-global variables. Also I am confused as to how many style guides we have now..

File: Changelog.md (hunk 5 above)
Comment (Marenz) [reply Changelog.md]: While this is true for the changelog in the code we never use that camel case spelling. Seems generally inconsistent

Comment (hrkrshnn) [reply Changelog.md]: The docs use the same "spelling" https://solidity.readthedocs.io/en/v0.6.9/natspec-format.html

File: libsolidity/interface/Natspec.h (hunk 4 above)
Comment (Marenz) [reply libsolidity/interface/Natspec.h]: oooh I looked at the solidity style guide :laughing:

**Comment by Marenz** [comment]
Anything missing here? From my side it seems ready :)

File: libsolidity/interface/Natspec.h [hunk 9]
```diff
@@ -40,6 +40,8 @@ struct DocTag;
 class Natspec
 {
 public:
+	static unsigned int constexpr NATSPEC_VERSION = 1;
```
Comment (hrkrshnn) [inline libsolidity/interface/Natspec.h]: This might need a replacement. Maybe `c_natspecVersion`?

Comment (hrkrshnn) [reply libsolidity/interface/Natspec.h]: @Marenz Sorry. It should be `s_` for static. https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md#4-variable-prefixes

Or maybe we need a new style for constants.

Comment (Marenz) [reply libsolidity/interface/Natspec.h]: I've also seen `libevmasm/CommonSubexpressionEliminator.h:	static int const c_invalidPosition = -0x7fffffff` so there is at least a precendent for this

**Review by hrkrshnn — DISMISSED** [review]
LGTM!

File: Changelog.md [hunk 10]
```diff
@@ -8,6 +8,7 @@ Compiler Features:
  * Commandline Interface: Re-group help screen.
  * Output compilation error codes in standard-json and when using ``--error-codes``.
  * Yul: Raise warning for switch statements that only have a default and no other cases.
+ * NatSpec: Add fields "kind" and "version" to the JSON output.
```
Comment (chriseth) [inline Changelog.md]: Wrong changelog.

Comment (Marenz) [reply Changelog.md]: What about it is wrong? That it is missing "documentation"? I figured the "NatSpec:" part would clarify that.

Comment (hrkrshnn) [reply Changelog.md]: @Marenz We already released 0.6.10. This should go in the unreleased section. I didn't realize this either ;)

Comment (Marenz) [reply Changelog.md]: ooh, will fix, thank you

**Review by chriseth — APPROVED** [review]

**Pull request merged**